### ISSUE TITLE
support $slices to explicitly include unit

### ISSUE THREAD

**Issue by betaboon** [lolmaus/breakpoint-slicer#30]
Hello,

I am in favor of not assuming `px` as the desired unit for media-queries and to let the user explicitly provide the unit with custom `$slices`.

Example:
```scss
$slices: (
  small: 0px,
  medium: 400px,
  large: 800px,
);
```

instead of:
```scss
$slices: (
  small: 0,
  medium: 400,
  large: 800,
);
```

would you be open for this?
If so, I'm happy to provide a PR.

**Comment by lolmaus** [comment]
I don't mind doing that. Seems to be useful for passing pre-existing variables:

```
$slices: (
  small: 0px,
  large: $third-party-breakpoint-in-px,
);
```

Can you make it support both `px` and unitless values, in order not to make a breaking change and keed the benefit of shorter syntax?

**Comment by lolmaus** [comment]
Please note that `breakpoint-slicer` is tightly coupled to `px`.

I. e. it's not gonna be possible to use other units because only `px` allow defining gaples, non-overlapping slices:

```
@media (max-width: 300px) {}
@media (min-width: 301px) {}
```

If you do it with e. g. `em`, you'll end up with a certain screen width that will either both or no queries applied.

Thus, if we allow the user to mess with units, we must check that the value is either in `px` or unitless, and throw an error otherwise.

**Comment by lolmaus** [comment]
Uhm, I'm now thinking that supporting both `px` and unitless is not a good idea. Unitless is convenient and short, but it loses the benefit of a small, strictly defined, strictly typed API.

Whatcha think?

**Comment by betaboon** [comment]
> Please note that breakpoint-slicer is tightly coupled to px.
> If you do it with e. g. em, you'll end up with a certain screen width that will either both or no queries applied.

what do you mean? the implementation is tightly coupled to px? or conceptually breakpoints are tightly coupled to px?
yes, there is a gap between `3em` and `3.1em` when talking about pixels. but that shouldnt matter.
I'll link some discussions on em-for-breakpoints.

> I'm now thinking that supporting both `px` and unitless is not a good idea.

I agree. either the use should be "forced" to explicitly define the unit, or a specific unit is always assumed.
I would prefer not assuming the unit in the library and relying on the user to supply the values with units.

my initial motiviation was this: in a library i am working on the breakpoint slices are configurable. now in the library-configuration all the settings include a unit and it just looks weird to have the settings for breakpoints not include some.
in general (following DDD) i do believe "numbers are meaningless without an approriate unit".

https://medium.com/zoosk-engineering/to-em-or-not-to-em-that-is-the-media-query-question-22f4a65e9747
https://zellwk.com/blog/media-query-units/

**Comment by lolmaus** [comment]
> what do you mean? the implementation is tightly coupled to px? or conceptually breakpoints are tightly coupled to px?

Breakpoints are tightly coupled to `px`.

Both your articles resolve the problem of gaps by producing *overlapping* queries which contain same CSS styles which overwrite each other:

```scss
.foo {
  color: red;

  @media (min-width: 10em) {
    color: blue;
  }

  @media (min-width: 20em) {
    color: green;
  }
}
```

This is OK when you need to sligtly tweak same rules. But if you need to apply entirely different layouts for different screen sizes, this approach becomes a mess:

```scss
.foo {
  float: left;

  @media (min-width: 10em) {
    float: none; // I don't want to be doing this!
    display: flex;
  }

  @media (min-width: 20em) {
    display: block; // Are you sure the original `display` was `block`? Or was it `inline`?
    position: absolute;
  }
}
```

In my work practice I have implemented many very elaborate responsive layouts, and having to stick with this approach would've been a nightmare. 

Instead, I use non-overlapping queries:

```scss
.foo {
  @media (max-width: 200px) {
    float: left;
  }

  @media (min-width: 201px) and (max-width: 400px) {
    display: flex;
  }

  @media (min-width: 401px) {
    position: absolute;
  }
}
```

Very clean, styles from different layouts no longer leak into each other! That's one of the reasons why Breakpoint Slicer exists.

But this can be done only with `px`.

Using the same `em` value in two queries like this:

```scss
.foo {
  @media (max-width: 10em) {
    float: left;
  }

  @media (min-width: 10em) {
    display: flex;
  }
}
```

...produces an overlap at which both queries apply, resulting in a faulty layout.

You could try using fractions of `em` but that's not reliable. It may still produce an overlap, or it may create a gap where neither layout applies. Also, I suspect browser settings such as the zoom level might affect this non-deterministically (I've seen a few times how zooming in produces off-by-one-pixel glitches caused by the rounding of non-px units).

The worst part is that you can't test it programmatically. You have to check by hand and you better dial down your mouse speed since you need to catch an exact viewport width with 0px error margin.

I want Breakpoint Slicer to produce reliable, continuous media query slices with neither overlaps nor gaps. It can only be achieved with `px`.

Internally, continuous non-overlapping slices are produced with [simple logic](https://github.com/lolmaus/breakpoint-slicer/blob/gen-3/index.scss#L324) of `$right-side-breakpoint - 1px`.

***

> I would prefer not assuming the unit in the library and relying on the user to supply the values with units.

Okay, let's stop supporting unitless values and start supporting `px` values only. Please file a PR!

***

> all the settings include a unit

I agree. I used unitless values because I was copying from my [ember-element-query](https://github.com/lolmaus/ember-element-query) JS library. There I had to use unitless values because JS is not capable of math with CSS units.

**Comment by betaboon** [comment]
thank you for this clarification. I'm now sold on your position of not supporting arbitrary units :)

when i find the time i will try to provide a PR to support `px` values only.

thank you for your insight

**Comment by lolmaus** [comment]
Fixed in #31, released in `3.0.0-beta.1`. 🎉
